### PR TITLE
[WIP] Added support for credentials-binding-plugin

### PR DESCRIPTION
**Pull request by garethahealy**
**DONT MERGE**

Added support for using the openshift bearer token inside the bindings plugin.

However, when i run the test for the new bindings code, i get "Host connection refused". If i do the same thing, but outside of the bindings, it works.

@jglick as you are one of the committers for the bindings plugins. any ideas what i am doing wrong?

### CODE REVIEW

**Comment by jglick** [comment]
I have no idea what OpenShift bearer tokens are, so I doubt I could be of assistance here.

File: src/test/java/org/jenkinsci/plugins/kubernetes/credentials/OpenShiftBearerTokenCredentialBindingTest.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+public class OpenShiftBearerTokenCredentialBindingTest {
+
+    @Rule
+    public JenkinsRule r = new JenkinsRule();
+
+    @ClassRule
+    public static BuildWatcher buildWatcher = new BuildWatcher();
+
+    @Rule
+    public RestartableJenkinsRule story = new RestartableJenkinsRule();
```
Comment (maxlaverse) [inline src/test/java/org/jenkinsci/plugins/kubernetes/credentials/OpenShiftBearerTokenCredentialBindingTest.java]: Does the implementation for the Bindings plugin require to use `RestartableJenkinsRule` ? 

If not, could you please test the `restartable` feature in another class ?

File: src/test/java/org/jenkinsci/plugins/kubernetes/credentials/OpenShiftBearerTokenCredentialBindingTest.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+import java.util.Collections;
+
+import static junit.framework.TestCase.assertTrue;
+import static org.codehaus.groovy.runtime.DefaultGroovyMethods.grep;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+
+public class OpenShiftBearerTokenCredentialBindingTest {
+
+    @Rule
```
Comment (maxlaverse) [inline src/test/java/org/jenkinsci/plugins/kubernetes/credentials/OpenShiftBearerTokenCredentialBindingTest.java]: I don't know anything about `RestartableJenkinsRule`, but I believe it replaces the `JenkinsRule` and therefore, this one can be removed

File: src/test/java/org/jenkinsci/plugins/kubernetes/credentials/OpenShiftBearerTokenCredentialBindingTest.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    public static BuildWatcher buildWatcher = new BuildWatcher();
+
+    @Rule
+    public RestartableJenkinsRule story = new RestartableJenkinsRule();
+
+    private Server server;
+    private String serverUri;
+
+    @Before
+    public void prepareFakeOAuthServer() throws Exception {
```
Comment (maxlaverse) [inline src/test/java/org/jenkinsci/plugins/kubernetes/credentials/OpenShiftBearerTokenCredentialBindingTest.java]: Again, I don't know anything about `RestartableJenkinsRule`, but if you add some log traces you can see that `@After` is executed before the story actually runs. This means your mock server will be stopped before your Workflow job executes. I believe that's why you tried to save `serverUri` in a shared variable

Not sure if it's the right thing to do, but starting the mock server as a first step solves the loading issue.
```
        story.addStep(new Statement() {
            @Override public void evaluate() throws Throwable {
                ServletContextHandler context = new ServletContextHandler();
                context.setContextPath("/");
                context.addServlet(new ServletHolder(new MockHttpServlet()), "/*");

                server = new Server(0);
                server.setHandler(context);
                server.start();
            }
        });
```

File: src/test/java/org/jenkinsci/plugins/kubernetes/credentials/OpenShiftBearerTokenCredentialBindingTest.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        server.stop();
+    }
+
+    @Test
+    public void basics() throws IOException {
+        final String credentialsId = "creds";
+        final String username = "bob";
+        final String password = "s$$cr3t";
+
+        OpenShiftBearerTokenCredentialImpl t = new OpenShiftBearerTokenCredentialImpl(CredentialsScope.GLOBAL, credentialsId, "sample", username, password);
```
Comment (maxlaverse) [inline src/test/java/org/jenkinsci/plugins/kubernetes/credentials/OpenShiftBearerTokenCredentialBindingTest.java]: Those two lines are most probably left-over of a debugging session

File: src/test/java/org/jenkinsci/plugins/kubernetes/credentials/OpenShiftBearerTokenCredentialBindingTest.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+                WorkflowRun b = p.getBuildByNumber(1);
+                assertNotNull(b);
+                assertEquals(Collections.<String>emptySet(), grep(b.getRootDir(), password));
+
+                SemaphoreStep.success("basics/1", null);
+                story.j.waitForCompletion(b);
+                story.j.assertBuildStatusSuccess(b);
+                story.j.assertLogNotContains(password, b);
+
+                FilePath script = story.j.jenkins.getWorkspaceFor(p).child("script");
```
Comment (maxlaverse) [inline src/test/java/org/jenkinsci/plugins/kubernetes/credentials/OpenShiftBearerTokenCredentialBindingTest.java]: Again, I don't know about `RestartableJenkinsRule` but I'll try to help a bit.
What should write this `script` file ? If it Jenkins that stores some local state when interrupting pipeline for a restart, it doesn't sound like something we should test as part of this plugin.

If your goal is to test if the pipeline behaves properly when being resumed from the `semaphore 'basics'` within `withCredentials`, wouldn't this tests be enough:
1. Execute the pipeline and interrupt it within `withCredentials` when to token is already available
2. Resume the pipeline and execute a script that actually stores the token into a file
3. In a third step, check that the file contains the expected token.

This definitely goes green when running the tests but maybe that's not what you want to achieve

File: src/test/java/org/jenkinsci/plugins/kubernetes/credentials/OpenShiftBearerTokenCredentialBindingTest.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        });
+
+        story.addStep(new Statement() {
+            @Override public void evaluate() throws Throwable {
+                WorkflowJob p = story.j.jenkins.getItemByFullName("p", WorkflowJob.class);
+                assertNotNull(p);
+
+                WorkflowRun b = p.getBuildByNumber(1);
+                assertNotNull(b);
+                assertEquals(Collections.<String>emptySet(), grep(b.getRootDir(), password));
```
Comment (maxlaverse) [inline src/test/java/org/jenkinsci/plugins/kubernetes/credentials/OpenShiftBearerTokenCredentialBindingTest.java]: `grep` complains about `b.getRootDir()` being a folder and not a file

**Comment by maxlaverse** [comment]
Closing as stale. If you want to get this merged, please make the requested changes and reopen.

**Pull request closed without merging**